### PR TITLE
Redis List API support

### LINKED ISSUE #201: [NEW]Fill the API/functionality gap of KVDK::List and Redis::List.
**The problem/use-case that the feature addresses**

The Redis::List commands require KVDK::List provide more functions.

**Description of the feature**

E.g.
Redis::List::LLEN ---> KVDK return the length of the list
Redis::List::LPOS ---> KVDK returns the index of matching elements inside a KVDK::List
etc.

### PR DESCRIPTION
**Pull request by ZiyanShi**
### What problem does this PR solve?

Issue Number: close #201

Problem Summary:

### What is changed and how it works?

What's Changed:
1. Reimplement persistent doubly-linked list. 
1.1 By removing persistent auxiliary head and tail node, performance for adding node to front and back of list is greatly improved.
1.2 Implement ListBuilder class to help rebuild the List during reboot.
2. Implemented some APIs for Redis List data type.

### Check List 

Tests 

- Unit test

Side effects
Improve List(Originally Queue) performance by about 200%.

### CODE REVIEW

File: engine/generic_list.hpp [hunk 1]
```diff
@@ -0,0 +1,768 @@
...
+ public:
+  explicit AddressTranslator(char* base) : pmem_base{base} {}
+  AddressTranslator() = default;
+  AddressTranslator(AddressTranslator const&) = default;
+  AddressTranslator(AddressTranslator&&) = default;
+  AddressTranslator& operator=(AddressTranslator const&) = default;
+  AddressTranslator& operator=(AddressTranslator&&) = default;
+  ~AddressTranslator() = default;
+
+  RecordPointer address_of(PMemOffsetType off) const {
```
Comment (JiayuZzz) [inline engine/generic_list.hpp:46]: AddressOf?

File: engine/skiplist.cpp [hunk 2]
```diff
@@ -361,6 +361,7 @@ bool Skiplist::Update(const StringView& internal_key, const StringView& value,
       pmem_allocator_->offset2addr(space_to_write.offset), space_to_write.size,
       timestamp, record_type, updated_offset, prev_offset, next_offset,
       internal_key, value, expired_time);
+  expire_time = expired_time;
```
Comment (sherry-1001) [inline engine/skiplist.cpp:364]: remove

File: benchmark/bench.cpp [hunk 3]
```diff
@@ -179,6 +192,7 @@ void DBWrite(int tid) {
 
     // generate key
     std::uint64_t num = generate_key(tid);
+    std::uint64_t cid = generate_cid(tid);
```
[the remark this answers is not in the document]
Comment (ZiyanShi) [reply benchmark/bench.cpp]: done

File: engine/data_record.hpp [hunk 4]
```diff
@@ -188,7 +202,14 @@ struct DLRecord {
   PMemOffsetType older_version_offset;
   PMemOffsetType prev;
   PMemOffsetType next;
-  ExpiredTimeType expired_time;
+  union {
```
[the remark this answers is not in the document]
Comment (ZiyanShi) [reply engine/data_record.hpp:205]: for collection head, the id is stored in value field.

File: engine/data_record.hpp [hunk 5]
```diff
@@ -223,13 +244,63 @@ struct DLRecord {
...
+    _mm_clwb(&prev);
+  }
+
+  void PersistExpireTimeCLWB(ExpiredTimeType time) {
+    kvdk_assert(entry.meta.type & ExpirableRecordType, "");
+    expired_time = time;
+    _mm_clwb(&expired_time);
+  }
+
+  ExpiredTimeType ExpireTime() {
```
[the remark this answers is not in the document]
Comment (ZiyanShi) [reply engine/data_record.hpp:299]: const is added.

File: engine/generic_list.hpp [hunk 6]
```diff
@@ -0,0 +1,777 @@
...
+ public:
+  explicit AddressTranslator(char* base) : pmem_base{base} {}
+  AddressTranslator() = default;
+  AddressTranslator(AddressTranslator const&) = default;
+  AddressTranslator(AddressTranslator&&) = default;
+  AddressTranslator& operator=(AddressTranslator const&) = default;
+  AddressTranslator& operator=(AddressTranslator&&) = default;
+  ~AddressTranslator() = default;
+
+  RecordPointer addressOf(PMemOffsetType off) const {
```
[the remark this answers is not in the document]
Comment (ZiyanShi) [reply engine/generic_list.hpp:46]: Don't make names too long.

File: engine/generic_list.hpp [hunk 7]
```diff
@@ -0,0 +1,777 @@
...
+    sz = n;
+  }
+
+  LockType* Mutex() { return &mu; }
+
+  std::unique_lock<LockType> AcquireLock() {
+    return std::unique_lock<LockType>(mu);
+  }
+
+  void ExpireAt(ExpiredTimeType time) final {
```
[the remark this answers is not in the document]
Comment (ZiyanShi) [reply engine/generic_list.hpp:222]: Follow Redis convention

File: engine/generic_list.hpp [hunk 8]
```diff
@@ -0,0 +1,777 @@
...
+      return iter;
+    } else {
+      auto iter = Back();
+      while (index != -1 && iter != Head()) {
+        --iter;
+        ++index;
+      }
+      return iter;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ZiyanShi) [reply engine/generic_list.hpp:253]: how?

File: engine/generic_list.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (sherry-1001) [reply engine/generic_list.hpp:222]: please give a example。 Internal used function no need to follow redis.

File: engine/data_record.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (sherry-1001) [reply engine/data_record.hpp:205]: move to next PR: duplicate key.

File: engine/kv_engine.cpp [hunk 9]
```diff
@@ -1693,15 +1667,42 @@ Status KVEngine::Expire(const StringView str, TTLTimeType ttl_time) {
...
-      Status::NotSupported;
+    case HashIndexType::UnorderedCollection: {
+      UnorderedCollection* pcoll;
+      auto guard = hash_table_->AcquireLock(str);
+      s = FindCollection(str, &pcoll, RecordType::DlistRecord);
+      if (s != Status::Ok) {
+        return s;
+      }
+      /// TODO: put these unregister and delete work in findKey()
+      pcoll->ExpireAt(expired_time);
```
[the remark this answers is not in the document]
Comment (ZiyanShi) [reply engine/kv_engine.cpp:1678]: If expired, find should have cleaned out the expired key and returned Status::NotFound.
I am working on the PR to provide a uniform API to find and remove expired key if necessary.

File: engine/kv_engine.cpp [hunk 10]
```diff
@@ -1693,15 +1667,42 @@ Status KVEngine::Expire(const StringView str, TTLTimeType ttl_time) {
...
+        unregisterCollection<UnorderedCollection>(str);
+        /// TODO: Also delete from PMem
+        return Status::NotFound;
+      }
+      return Status::Ok;
+    }
+    case HashIndexType::List: {
+      std::unique_lock<std::recursive_mutex> guard_list;
+      List* list;
+      s = listFind(str, &list, false, guard_list);
```
[the remark this answers is not in the document]
Comment (ZiyanShi) [reply engine/kv_engine.cpp:1689]: Peifeng worries that if we don't lock the List but the Slot, other keys in the same Slot may be blocked.

File: engine/skiplist.cpp (hunk 2 above)
Comment (ZiyanShi) [reply engine/skiplist.cpp:364]: You gotta update the field in DRAM

File: engine/kv_engine.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (sherry-1001) [reply engine/kv_engine.cpp:1678]: no. If the collection is expired, why do you set `expired time`? So firstly check is expired. And Suggest internal API are `setexpired` and `getexpired`, it's more clear.

File: engine/kv_engine.cpp [hunk 11]
```diff
@@ -1693,15 +1667,42 @@ Status KVEngine::Expire(const StringView str, TTLTimeType ttl_time) {
...
+      if (s != Status::Ok) {
+        return s;
+      }
+      list->ExpireAt(expired_time);
```
[the remark this answers is not in the document]
Comment (ZiyanShi) [reply engine/kv_engine.cpp:1693]: Let's do it in a uniformed API

File: engine/skiplist.hpp [hunk 12]
```diff
@@ -193,8 +193,8 @@ class Skiplist : public Collection {
 
   SkiplistNode* header() { return header_; }
 
-  ExpiredTimeType GetExpiredTime() const override {
-    return header_->record->GetExpiredTime();
+  void ExpireAt(ExpiredTimeType time) final {
+    throw std::runtime_error{"Unimplemented!"};
```
[the remark this answers is not in the document]
Comment (ZiyanShi) [reply engine/skiplist.hpp:197]: Added an expire_time field in Collection. This field resides in DRAM to accelerate ExpireTime().
Every collection should maintain this field by ExpireAt() or other operations that modified the pmem record for the collection.

File: engine/generic_list.hpp [hunk 13]
```diff
@@ -0,0 +1,777 @@
...
+
+  template <typename ElemDeleter>
+  Iterator Erase(Iterator pos, ElemDeleter elem_deleter) {
+    kvdk_assert(pos != Head(), "Cannot erase Head()");
```
Comment (sherry-1001) [inline engine/generic_list.hpp:257]: can erase Tail() ?

Comment (ZiyanShi) [reply engine/generic_list.hpp:230]: Tail() is Head(). It's a dummy iterator. Please refer to comments.

**Review by ZiyanShi — COMMENTED** [review]
done

File: engine/simple_list.hpp [hunk 14]
```diff
@@ -0,0 +1,13 @@
+#pragma once
```
Comment (JiayuZzz) [inline engine/simple_list.hpp:1]: put these in alias.hpp?

File: engine/skiplist.cpp (hunk 2 above)
Comment (JiayuZzz) [inline engine/skiplist.cpp:364]: I think better do not touch expire in this pr

File: engine/skiplist.hpp [hunk 15]
```diff
@@ -193,8 +193,14 @@ class Skiplist : public Collection {
 
   SkiplistNode* header() { return header_; }
 
-  ExpiredTimeType GetExpiredTime() const override {
-    return header_->record->GetExpiredTime();
+  ExpiredTimeType GetExpireTime() const final { return expire_time; }
```
Comment (JiayuZzz) [inline engine/skiplist.hpp:196]: I think better do not touch expire in this pr

File: include/kvdk/collection.hpp [hunk 16]
```diff
@@ -4,41 +4,75 @@
...
-  virtual int64_t GetExpiredTime() const = 0;
+  virtual ExpiredTimeType GetExpireTime() const = 0;
+  virtual bool HasExpired() const = 0;
+  virtual void SetExpireTime(ExpiredTimeType) = 0;
 
   // Return internal representation of "key" in the collection
   // By default, we concat key with the collection id
   std::string InternalKey(const StringView& key) {
     return makeInternalKey(key, ID());
   }
```
Comment (JiayuZzz) [inline include/kvdk/collection.hpp:45]: do not touch collection in this PR

**Pull request merged**